File: EECE7205-Advanced-Algorithms/HW4/johnson.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <map>
#include <iomanip>
using namespace std;

vector<int> bf_dist;

bool bellmanFord(int src, int V, map<int, vector<pair<int, int>>>& G) {
    vector<int> d(V, INT_MAX);
    d[src] = 0;

    for (int i = 0; i < V - 1; ++i) {
        for (int u = 0; u < V; ++u) {
            if (G.find(u) != G.end()) {
                for (auto& e : G[u]) {
                    int v = e.first;
                    int w = e.second;
                    if (d[u] != INT_MAX && d[u] + w < d[v]) {
                        d[v] = d[u] + w;
                    }
                }
            }
        }
    }

    for (int u = 0; u < V; ++u) {
        if (G.find(u) != G.end()) {
            for (auto& e : G[u]) {
                int v = e.first;
                int w = e.second;
                if (d[u] != INT_MAX && d[u] + w < d[v]) {
                    cout << "Graph contains a negative weight cycle." << endl;
                    return false;
                }
            }
        }
    }
    bf_dist = d;
    return true;
}

vector<int> dijkstra(int src, int V, map<int, vector<pair<int, int>>>& G) {
    vector<int> d(V, INT_MAX);
    vector<bool> visited(V, false);
    vector<pair<int, int>> Q;

    d[src] = 0;
    Q.push_back({0, src});

    while (!Q.empty()) {
        int min_idx = 0;
        for (int i = 1; i < Q.size(); i++) {
            if (Q[i].first < Q[min_idx].first) {
                min_idx = i;
            }
        }

        int dist = Q[min_idx].first;
        int u = Q[min_idx].second;
        Q.erase(Q.begin() + min_idx);

        if (visited[u]) {
            continue;
        }
        visited[u] = true;

        if (G.find(u) != G.end()) {
            for (auto& e : G[u]) {
                int v = e.first;
                int w = e.second;

                if (!visited[v] && d[u] + w < d[v]) {
                    d[v] = d[u] + w;
                    Q.push_back({d[v], v});
                }
            }
        }
    }
    return d;
}

void johnson(int V, map<int, vector<pair<int, int>>>& G) {
    for (int u = 0; u < V; ++u) {
        G[V].emplace_back(u, 0);
    }

    if (!bellmanFord(V, V + 1, G)) {
        cout << "Negative weight cycle detected. Algorithm cannot proceed." << endl;
        return;
    }

    vector<int> h = bf_dist;
    map<int, vector<pair<int, int>>> new_G;

    for (int u = 0; u < V; ++u) {
        for (auto& [v, w] : G[u]) {
            // w'(u,v) = w(u,v) + h(u) - h(v)
            int new_w = w + h[u] - h[v];
            new_G[u].emplace_back(v, new_w);
        }
    }
    cout << endl;

    cout << "New Graph (Graph after adjusting edge weights to eliminate negative weights):" << endl;
    for (int u = 0; u < V; ++u) {
        for (int v = 0; v < V; ++v) {
            int w = INT_MAX;
            if (new_G.find(u) != new_G.end()) {
                for (const auto& e : new_G[u]) {
                    if (e.first == v) {
                        w = e.second;
                        break;
                    }
                }
            }
            cout << setw(5) << (w == INT_MAX ? "0" : to_string(w));
        }
        cout << endl;
    }

    vector<vector<int>> all_pairs_distances(V, vector<int>(V, INT_MAX));
    for (int u = 0; u < V; ++u) {
        vector<int> distances = dijkstra(u, V, new_G);

        for (int v = 0; v < V; ++v) {
            if (distances[v] < INT_MAX) {
                all_pairs_distances[u][v] = distances[v] + h[v] - h[u];
            }
        }
    }

    cout << "Original Graph Shortest Paths Between All Pairs (After Conversion from Reweighted Graph):" << endl;
    for (int u = 0; u < V; ++u) {
        for (int v = 0; v < V; ++v) {
            cout << setw(5) << (all_pairs_distances[u][v] == INT_MAX ? "inf" : to_string(all_pairs_distances[u][v]));
        }
        cout << endl;
    }
}

int main() {
    int V = 4;
    vector<vector<int>> graph = {
        {0, -5, 2, 3},
        {0,  0, 4, 0},
        {0,  0, 0, 1},
        {0,  0, 0, 0}
    };

    map<int, vector<pair<int, int>>> G;
    for (int u = 0; u < V; ++u) {
        for (int v = 0; v < V; ++v) {
            if (graph[u][v] != 0) {
                G[u].emplace_back(v, graph[u][v]);
            }
        }
    }

    johnson(V, G);
    return 0;
}

File: EECE7205-Advanced-Algorithms/HW4/johnson.py
import networkx as nx
import matplotlib.pyplot as plt
from collections import defaultdict
import heapq

INF = float('inf')

def visualize_full_process(original_graph, reweighted_graph, potentials, V):
    """Visualizes the full process: Original Graph, Bellman-Ford Potentials, and Reweighted Graph."""
    fig, axes = plt.subplots(1, 2, figsize=(18, 6))

    # Plot Original Graph
    G1 = nx.DiGraph()
    for u in range(V):
        if u in original_graph:
            for v, w in original_graph[u]:
                G1.add_edge(u, v, weight=w)

    pos = nx.spring_layout(G1, seed=42)
    edge_labels = nx.get_edge_attributes(G1, 'weight')
    nx.draw(G1, pos, with_labels=True, ax=axes[0], node_color='lightblue', node_size=2000, font_size=12, font_weight='bold')
    nx.draw_networkx_edge_labels(G1, pos, edge_labels=edge_labels, font_size=10, ax=axes[0])
    axes[0].set_title("Original Graph with Negative Edges Weights")

    # Plot Reweighted Graph
    G3 = nx.DiGraph()
    for u in range(V):
        if u in reweighted_graph:
            for v, w in reweighted_graph[u]:
                G3.add_edge(u, v, weight=w)

    edge_labels_reweighted = nx.get_edge_attributes(G3, 'weight')
    nx.draw(G3, pos, with_labels=True, ax=axes[1], node_color='lightblue', node_size=2000, font_size=12, font_weight='bold')
    nx.draw_networkx_edge_labels(G3, pos, edge_labels=edge_labels_reweighted, font_size=10, ax=axes[1])
    axes[1].set_title("Transformed Graph with No Negative Edge Weights")

    plt.tight_layout()
    plt.show()

def bellman_ford(src, V, G):
    d = [INF] * V
    d[src] = 0

    for _ in range(V - 1):
        for u in range(V):
            if u in G:
                for v, w in G[u]:
                    if d[u] != INF and d[u] + w < d[v]:
                        d[v] = d[u] + w

    for u in range(V):
        if u in G:
            for v, w in G[u]:
                if d[u] != INF and d[u] + w < d[v]:
                    print("Graph contains a negative weight cycle.")
                    return False, []

    return True, d

def johnson(V, G):
    original_graph = G.copy()

    G[V] = [(u, 0) for u in range(V)]

    has_no_neg_cycle, h = bellman_ford(V, V + 1, G)
    if not has_no_neg_cycle:
        print("Negative weight cycle detected. Algorithm cannot proceed.")
        return

    del G[V]

    reweighted_graph = defaultdict(list)
    for u in range(V):
        for v, w in G[u]:
            new_w = w + h[u] - h[v]
            reweighted_graph[u].append((v, new_w))

    visualize_full_process(original_graph, reweighted_graph, h, V)

if __name__ == "__main__":
    V = 4
    graph = [
        [0, -5, 2, 3],
        [0,  0, 4, 0],
        [0,  0, 0, 1],
        [0,  0, 0, 0]
    ]

    G = defaultdict(list)
    for u in range(V):
        for v in range(V):
            if graph[u][v] != 0:
                G[u].append((v, graph[u][v]))

    johnson(V, G)